Zend/zend_compile.c

```diff
@@ -2867,8 +2865,33 @@ static void zend_compile_list_assign(
 
 		zend_verify_list_assign_target(var_ast, old_style);
 
-		zend_emit_op(&fetch_result, ZEND_FETCH_LIST, expr_node, &dim_node);
-		zend_emit_assign_znode(var_ast, &fetch_result);
+		opline = zend_emit_op(&fetch_result, ZEND_FETCH_LIST, expr_node, &dim_node);
+		if (elem_ast->attr) {
+			opline->extended_value = ZEND_LIST_MAKE_WRITABLE;
+			prev_op = (opline - 1);
+			do {
+				if (prev_op->opcode == ZEND_FETCH_LIST) {
+					if (prev_op->op1.var == opline->op1.var) {
+						continue;
+					} else if (prev_op->result.var == opline->op1.var) {
+						if (prev_op->extended_value == ZEND_LIST_MAKE_WRITABLE) {
+							break;
+						}
+						prev_op->extended_value = ZEND_LIST_MAKE_WRITABLE;
+						opline = prev_op;
+						continue;
+					}
+					break;
+				} else if (prev_op->opcode == ZEND_ASSIGN_REF || prev_op->opcode == ZEND_ASSIGN) {
```

Comment:
It may not be perfect, but I changed it to be a pre-look that for each elem it'd look down the list seeing if there's references so we can set flag accordingly.

As to the next issue with indirect, I'm not sure how to get a case in which it fails.
```
$a = [
    1 => "one",
    2 => "two",
];

[[1 => &$b, $a[] = 2 => $c]] = $a;
var_dump($a);
```

outputs
```
PHP Notice:  Undefined offset: 0 in /home/dwalker/src/php/e.php on line 8
PHP Notice:  Undefined offset: 1 in /home/dwalker/src/php/e.php on line 8
PHP Notice:  Undefined offset: 2 in /home/dwalker/src/php/e.php on line 8
array(4) {
  [1]=>
  string(3) "one"
  [2]=>
  string(3) "two"
  [0]=>
  array(1) {
    [1]=>
    &NULL
  }
  [3]=>
  int(2)
}
```

Which is what I'm kind of expecting to see.  The `$a` main array remains after the run.  There's a `0`'th element that is an array with the reference to `NULL`, and the odd key of `$a[] = 2` does give a new element in `$a` with the value 2.  If there some trick to make the indirect fail better to try and understand it better.